File: LongestSubStr.java
// Time Complexity : O(n). n is the length of the string.
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        if(s==null || s.length()==0) return 0;
        
        int low=0,max=0,slow=0,fast=0;
        HashMap<Character,Integer> map = new HashMap();
        
        while(fast<s.length()){
            char ch = s.charAt(fast);
            
            if(map.containsKey(ch)){
                slow = Math.max(slow,map.get(ch));    
            }
            map.put(ch,fast+1);
            max = Math.max(max,fast-slow+1);
            fast++;
        }
        return max;
    }
}

File: LongestSubstr.py
"""
Given a string s, find the length of the longest substring without repeating characters.

"""

# Time Complexity : O(n)
# Space Complexity : O(n)
# Did this code successfully run on leetcode : Yes
# Any problem you faced while coding this : No

from typing import List

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        
        # base
        if not s or len(s) == 0:
            return 0
    
        # logic
       
        slow = 0
        max1 = 0
        map1 = {}
        
        for i in range(len(s)):
            c = s[i]
            if c in map1:
                slow = max(slow, map1[c])
                
            map1[c] = i+1
            max1 = max(max1, i - slow + 1)
            
        return max1
        